File: python/src/hearth/apikeyenvelope.py
"""The wire format for handing an API key to a recipient that published an
X25519 public key — typically a confidential VM (Intel TDX) that generated
the key inside the TD and bound it into its attestation report.

The envelope is a thin, self-describing frame around :mod:`hearth.hpke`::

    offset size field
    0      4    "HKE1"                      format magic and version
    4      2    kem_id                      0x0020, DHKEM(X25519, HKDF-SHA256)
    6      2    kdf_id                      0x0001, HKDF-SHA256
    8      2    aead_id                     0x0003 by default, ChaCha20-Poly1305
    10     8    fingerprint                 SHA-256(recipient public key)[0..8]
    18     2    metadata_len
    20     m    metadata                    key id and expiry, see Metadata
    20+m   32   enc                         the encapsulated key
    52+m   48   ciphertext                  32-byte API key + 16-byte tag

Everything before ``enc`` is passed to the AEAD as additional authenticated
data, so the suite ids, the recipient fingerprint and the metadata are all
covered by the tag: an envelope cannot be re-labelled with a different key id
or expiry, and the ``info`` string pins it to this protocol so it cannot be
replayed into another one.

The fingerprint is a routing hint, not a security control — the recipient
uses it to reject an envelope sealed to a previous boot's key with a clear
error instead of an authentication failure.

**What this does not give you.** HPKE base mode does not authenticate the
sender, and an envelope stays decryptable as long as the recipient's private
key lives: authorize the delivery request at the transport layer, keep the
recipient keypair ephemeral per boot, and set ``Metadata.not_after``. And none
of it means anything until the caller has verified the attestation quote and
checked that the recipient's public key is the one bound into ``REPORTDATA``.
"""

from __future__ import annotations

import hashlib
import secrets
from dataclasses import dataclass
from datetime import UTC, datetime

from . import hpke, x25519

# API keys are exactly this many characters.
API_KEY_LENGTH = 32

# ChaCha20-Poly1305: no reliance on the TD having usable AES-NI.
DEFAULT_SUITE = hpke.Suite.X25519_SHA256_CHACHA20POLY1305

# Bytes of SHA-256(public key) carried in the header.
FINGERPRINT_BYTES = 8

# The HPKE info string. Changing it breaks compatibility, by design.
_INFO = b"hearth-chain/api-key-hpke/v1"

_MAGIC = b"HKE1"
_HEADER_FIXED_BYTES = 20
_MAX_METADATA_BYTES = 0xFFFF

_ALPHANUMERIC = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
# Largest multiple of 62 that fits in a byte; above it, resample (no modulo bias).
_SAMPLE_LIMIT = (256 // len(_ALPHANUMERIC)) * len(_ALPHANUMERIC)


@dataclass(frozen=True)
class Metadata:
    """What the envelope claims about the key it carries, authenticated by
    the AEAD tag.

    ``key_id`` is an identifier for the API key, 1..255 bytes of UTF-8; it
    lets the recipient tell which key it received without logging the key
    itself. ``not_after`` is when the key stops being valid, or ``None`` for
    no expiry. Truncated to whole seconds on the wire.
    """

    key_id: str
    not_after: datetime | None = None

    def __post_init__(self) -> None:
        length = len(self.key_id.encode("utf-8"))
        if length < 1 or length > 255:
            raise ValueError(f"key id must be 1..255 bytes of UTF-8, was {length}")

    def _encode(self) -> bytes:
        id_bytes = self.key_id.encode("utf-8")
        epoch_seconds = 0 if self.not_after is None else int(self.not_after.timestamp())
        return bytes([len(id_bytes)]) + id_bytes + epoch_seconds.to_bytes(8, "big", signed=True)

    @staticmethod
    def _decode(encoded: bytes) -> Metadata:
        if len(encoded) < 1 + 8:
            raise ValueError("truncated envelope metadata")
        length = encoded[0]
        if len(encoded) != 1 + length + 8:
            raise ValueError("envelope metadata length mismatch")
        key_id = encoded[1 : 1 + length].decode("utf-8")
        epoch_seconds = int.from_bytes(encoded[1 + length :], "big", signed=True)
        not_after = None if epoch_seconds == 0 else datetime.fromtimestamp(epoch_seconds, tz=UTC)
        return Metadata(key_id, not_after)


@dataclass
class Opened:
    """A decrypted envelope. Call :meth:`wipe` once the key has been used —
    the point of returning a ``bytearray`` is that it can be cleared, which a
    ``str`` cannot."""

    api_key: bytearray
    metadata: Metadata

    def wipe(self) -> None:
        for i in range(len(self.api_key)):
            self.api_key[i] = 0


def seal(
    recipient_public_key: bytes,
    api_key: str | bytearray,
    metadata: Metadata,
    suite: hpke.Suite = DEFAULT_SUITE,
) -> bytes:
    """Seal an API key to ``recipient_public_key``.

    The caller must already have verified that this public key belongs to
    the enclave it expects; this function cannot check that.

    Raises ValueError if the API key is not :data:`API_KEY_LENGTH`
    alphanumeric characters.
    """
    key_str = api_key.decode("ascii") if isinstance(api_key, (bytes, bytearray)) else api_key
    _validate_api_key(key_str)
    header = _header(suite, fingerprint(recipient_public_key), metadata._encode())
    plaintext = key_str.encode("ascii")
    sealed = hpke.seal(suite, recipient_public_key, _INFO, header, plaintext)
    return header + sealed.enc + sealed.ciphertext


def open_(recipient_secret_key: bytes, envelope: bytes, now: datetime | None = None) -> Opened:
    """Open an envelope, rejecting one that has expired.

    ``now`` is the instant to judge ``Metadata.not_after`` against (defaults
    to the current time).

    Raises ValueError if the envelope is malformed, sealed to a different
    recipient key, not authentic, or expired.
    """
    if now is None:
        now = datetime.now(UTC)
    if len(envelope) < _HEADER_FIXED_BYTES:
        raise ValueError("truncated envelope")
    if envelope[:4] != _MAGIC:
        raise ValueError("not an API key envelope")

    kem_id = int.from_bytes(envelope[4:6], "big")
    kdf_id = int.from_bytes(envelope[6:8], "big")
    aead_id = int.from_bytes(envelope[8:10], "big")
    if kem_id != hpke.KEM_ID or kdf_id != hpke.KDF_ID:
        raise ValueError(f"unsupported HPKE suite: kem=0x{kem_id:04x} kdf=0x{kdf_id:04x}")
    suite = hpke.Suite.from_aead_id(aead_id)

    env_fingerprint = envelope[10:18]
    metadata_length = int.from_bytes(envelope[18:20], "big")

    ciphertext_length = API_KEY_LENGTH + hpke.TAG_BYTES
    expected = _HEADER_FIXED_BYTES + metadata_length + hpke.ENC_BYTES + ciphertext_length
    if len(envelope) != expected:
        raise ValueError(f"envelope length mismatch: expected {expected} bytes, got {len(envelope)}")

    offset = _HEADER_FIXED_BYTES
    metadata_bytes = envelope[offset : offset + metadata_length]
    offset += metadata_length
    enc = envelope[offset : offset + hpke.ENC_BYTES]
    offset += hpke.ENC_BYTES
    ciphertext = envelope[offset : offset + ciphertext_length]

    recipient_public_key = x25519.public_key(recipient_secret_key)
    if env_fingerprint != fingerprint(recipient_public_key):
        raise ValueError("envelope is sealed to a different recipient key")

    # Everything before enc is the AAD, so the suite ids, fingerprint and
    # metadata are all covered by the tag.
    header = envelope[: _HEADER_FIXED_BYTES + metadata_length]
    plaintext = hpke.open_(suite, recipient_secret_key, enc, _INFO, header, ciphertext)

    api_key = plaintext.decode("ascii")
    _validate_api_key(api_key)
    metadata = Metadata._decode(metadata_bytes)
    if metadata.not_after is not None and now >= metadata.not_after:
        raise ValueError(f"envelope expired at {metadata.not_after}")
    return Opened(bytearray(plaintext), metadata)


def fingerprint(public_key: bytes) -> bytes:
    """SHA-256(public key) truncated to :data:`FINGERPRINT_BYTES` bytes."""
    if len(public_key) != x25519.KEY_BYTES:
        raise ValueError(f"public key must be {x25519.KEY_BYTES} bytes")
    return hashlib.sha256(public_key).digest()[:FINGERPRINT_BYTES]


def random_api_key() -> str:
    """A fresh :data:`API_KEY_LENGTH`-character alphanumeric API key, uniform
    over the 62-character alphabet (rejection sampling — ``% 62`` on a random
    byte would favour the first 8 characters)."""
    chars: list[str] = []
    while len(chars) < API_KEY_LENGTH:
        for b in secrets.token_bytes(API_KEY_LENGTH):
            if len(chars) >= API_KEY_LENGTH:
                break
            if b < _SAMPLE_LIMIT:
                chars.append(_ALPHANUMERIC[b % len(_ALPHANUMERIC)])
    return "".join(chars)


def _header(suite: hpke.Suite, fp: bytes, metadata: bytes) -> bytes:
    if len(metadata) > _MAX_METADATA_BYTES:
        raise ValueError(f"envelope metadata too long: {len(metadata)}")
    return (
        _MAGIC
        + hpke.KEM_ID.to_bytes(2, "big")
        + hpke.KDF_ID.to_bytes(2, "big")
        + suite.aead_id.to_bytes(2, "big")
        + fp
        + len(metadata).to_bytes(2, "big")
        + metadata
    )


def _validate_api_key(api_key: str) -> None:
    if len(api_key) != API_KEY_LENGTH:
        raise ValueError(f"API key must be {API_KEY_LENGTH} characters")
    if not all(c.isalnum() and c.isascii() for c in api_key):
        raise ValueError("API key must be alphanumeric")

File: go/hearth/apikeyenvelope.go
package hearth

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"time"
)

// ApiKeyEnvelope is the wire format for handing an API key to a recipient
// that published an X25519 public key — typically a confidential VM (Intel
// TDX) that generated the key inside the TD and bound it into its
// attestation report.
//
// The envelope is a thin, self-describing frame around HPKE (hpke.go):
//
//	offset size field
//	0      4    "HKE1"        format magic and version
//	4      2    kem_id        0x0020, DHKEM(X25519, HKDF-SHA256)
//	6      2    kdf_id        0x0001, HKDF-SHA256
//	8      2    aead_id       0x0003 by default, ChaCha20-Poly1305
//	10     8    fingerprint   SHA-256(recipient public key)[0..8]
//	18     2    metadata_len
//	20     m    metadata      key id and expiry, see ApiKeyMetadata
//	20+m   32   enc           the encapsulated key
//	52+m   48   ciphertext    32-byte API key + 16-byte tag
//
// Everything before enc is passed to the AEAD as additional authenticated
// data, so the suite ids, the recipient fingerprint and the metadata are all
// covered by the tag: an envelope cannot be re-labelled with a different key
// id or expiry, and the info string pins it to this protocol so it cannot be
// replayed into another one.
//
// The fingerprint is a routing hint, not a security control — the recipient
// uses it to reject an envelope sealed to a previous boot's key with a clear
// error instead of an authentication failure.
//
// What this does not give you: HPKE base mode does not authenticate the
// sender, and an envelope stays decryptable as long as the recipient's
// private key lives — authorize the delivery request at the transport
// layer, keep the recipient keypair ephemeral per boot, and set an expiry.
// And none of it means anything until the caller has verified the
// attestation quote and checked that the recipient's public key is the one
// bound into REPORTDATA.

// ApiKeyLength is the number of characters in an API key.
const ApiKeyLength = 32

// ApiKeyFingerprintBytes is the number of bytes of SHA-256(public key)
// carried in the header.
const ApiKeyFingerprintBytes = 8

// DefaultApiKeyEnvelopeSuite is ChaCha20-Poly1305: no reliance on the TD
// having usable AES-NI.
var DefaultApiKeyEnvelopeSuite = HpkeX25519Sha256ChaCha20Poly1305

// apiKeyEnvelopeInfo is the HPKE info string. Changing it breaks
// compatibility, by design.
var apiKeyEnvelopeInfo = []byte("hearth-chain/api-key-hpke/v1")

var apiKeyEnvelopeMagic = []byte("HKE1")

const apiKeyEnvelopeHeaderFixedBytes = 20
const apiKeyEnvelopeMaxMetadataBytes = 0xffff

const apiKeyAlphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"

// apiKeySampleLimit is the largest multiple of 62 that fits in a byte; above
// it, resample (no modulo bias).
const apiKeySampleLimit = (256 / len(apiKeyAlphabet)) * len(apiKeyAlphabet)

// ApiKeyMetadata is what the envelope claims about the key it carries,
// authenticated by the AEAD tag.
//
// KeyID identifies the API key, 1..255 bytes of UTF-8; it lets the recipient
// tell which key it received without logging the key itself. NotAfter is
// when the key stops being valid, or the zero Time for no expiry. Truncated
// to whole seconds on the wire.
type ApiKeyMetadata struct {
	KeyID    string
	NotAfter time.Time
}

// NewApiKeyMetadata builds metadata with no expiry.
func NewApiKeyMetadata(keyID string) (ApiKeyMetadata, error) {
	return NewApiKeyMetadataWithExpiry(keyID, time.Time{})
}

// NewApiKeyMetadataWithExpiry builds metadata that expires at notAfter (pass
// the zero Time for no expiry).
func NewApiKeyMetadataWithExpiry(keyID string, notAfter time.Time) (ApiKeyMetadata, error) {
	length := len(keyID)
	if length < 1 || length > 255 {
		return ApiKeyMetadata{}, fmt.Errorf("key id must be 1..255 bytes of UTF-8, was %d", length)
	}
	return ApiKeyMetadata{KeyID: keyID, NotAfter: notAfter}, nil
}

func (m ApiKeyMetadata) encode() []byte {
	id := []byte(m.KeyID)
	out := make([]byte, 0, 1+len(id)+8)
	out = append(out, byte(len(id)))
	out = append(out, id...)
	var epoch uint64
	if !m.NotAfter.IsZero() {
		epoch = uint64(m.NotAfter.Unix())
	}
	var epochBytes [8]byte
	binary.BigEndian.PutUint64(epochBytes[:], epoch)
	return append(out, epochBytes[:]...)
}

func decodeApiKeyMetadata(encoded []byte) (ApiKeyMetadata, error) {
	if len(encoded) < 1+8 {
		return ApiKeyMetadata{}, errors.New("truncated envelope metadata")
	}
	length := int(encoded[0])
	if len(encoded) != 1+length+8 {
		return ApiKeyMetadata{}, errors.New("envelope metadata length mismatch")
	}
	id := encoded[1 : 1+length]
	epoch := binary.BigEndian.Uint64(encoded[1+length:])
	m := ApiKeyMetadata{KeyID: string(id)}
	if epoch != 0 {
		m.NotAfter = time.Unix(int64(epoch), 0).UTC()
	}
	return m, nil
}

// OpenedApiKey is a decrypted envelope. Call Wipe once the key has been used.
type OpenedApiKey struct {
	ApiKey   []byte // ASCII characters, ApiKeyLength long
	Metadata ApiKeyMetadata
}

// Wipe zeroes the recovered API key bytes.
func (o OpenedApiKey) Wipe() {
	for i := range o.ApiKey {
		o.ApiKey[i] = 0
	}
}

// SealApiKey seals an API key with the default suite.
//
// The caller must already have verified that recipientPublicKey belongs to
// the enclave it expects; this function cannot check that.
func SealApiKey(recipientPublicKey, apiKey []byte, metadata ApiKeyMetadata) ([]byte, error) {
	return SealApiKeyWithSuite(recipientPublicKey, apiKey, metadata, DefaultApiKeyEnvelopeSuite)
}

// SealApiKeyWithSuite seals an API key to recipientPublicKey with the given
// suite.
//
// It returns an error if apiKey is not ApiKeyLength alphanumeric ASCII
// characters.
func SealApiKeyWithSuite(recipientPublicKey, apiKey []byte, metadata ApiKeyMetadata, suite HpkeSuite) ([]byte, error) {
	if err := validateApiKey(apiKey); err != nil {
		return nil, err
	}
	fingerprint, err := apiKeyFingerprint(recipientPublicKey)
	if err != nil {
		return nil, err
	}
	header, err := apiKeyEnvelopeHeader(suite, fingerprint, metadata.encode())
	if err != nil {
		return nil, err
	}
	sealed, err := HpkeSeal(suite, recipientPublicKey, apiKeyEnvelopeInfo, header, apiKey)
	if err != nil {
		return nil, err
	}
	out := make([]byte, 0, len(header)+len(sealed.Enc)+len(sealed.Ciphertext))
	out = append(out, header...)
	out = append(out, sealed.Enc...)
	out = append(out, sealed.Ciphertext...)
	return out, nil
}

// OpenApiKey opens an envelope, rejecting one that has expired as of now.
func OpenApiKey(recipientSecretKey, envelope []byte) (OpenedApiKey, error) {
	return OpenApiKeyAt(recipientSecretKey, envelope, time.Now())
}

// OpenApiKeyAt opens an envelope, judging ApiKeyMetadata.NotAfter against
// now.
//
// It returns an error if the envelope is malformed, sealed to a different
// recipient key, not authentic, or expired.
func OpenApiKeyAt(recipientSecretKey, envelope []byte, now time.Time) (OpenedApiKey, error) {
	if len(envelope) < apiKeyEnvelopeHeaderFixedBytes {
		return OpenedApiKey{}, errors.New("truncated envelope")
	}
	magic := envelope[0:4]
	if string(magic) != string(apiKeyEnvelopeMagic) {
		return OpenedApiKey{}, errors.New("not an API key envelope")
	}
	kemID := int(binary.BigEndian.Uint16(envelope[4:6]))
	kdfID := int(binary.BigEndian.Uint16(envelope[6:8]))
	aeadID := int(binary.BigEndian.Uint16(envelope[8:10]))
	if kemID != HpkeKemID || kdfID != HpkeKdfID {
		return OpenedApiKey{}, fmt.Errorf("unsupported HPKE suite: kem=0x%04x kdf=0x%04x", kemID, kdfID)
	}
	suite, err := HpkeSuiteFromAEADID(aeadID)
	if err != nil {
		return OpenedApiKey{}, err
	}

	fingerprint := envelope[10:18]
	metadataLength := int(binary.BigEndian.Uint16(envelope[18:20]))

	ciphertextLength := ApiKeyLength + HpkeTagBytes
	expected := apiKeyEnvelopeHeaderFixedBytes + metadataLength + HpkeEncBytes + ciphertextLength
	if len(envelope) != expected {
		return OpenedApiKey{}, fmt.Errorf("envelope length mismatch: expected %d bytes, got %d", expected, len(envelope))
	}
	metadataBytes := envelope[20 : 20+metadataLength]
	enc := envelope[20+metadataLength : 20+metadataLength+HpkeEncBytes]
	ciphertext := envelope[20+metadataLength+HpkeEncBytes:]

	recipientPublicKey, err := X25519PublicKey(recipientSecretKey)
	if err != nil {
		return OpenedApiKey{}, err
	}
	wantFingerprint, err := apiKeyFingerprint(recipientPublicKey)
	if err != nil {
		return OpenedApiKey{}, err
	}
	if string(fingerprint) != string(wantFingerprint) {
		return OpenedApiKey{}, errors.New("envelope is sealed to a different recipient key")
	}

	// Everything before enc is the AAD, so the suite ids, fingerprint and
	// metadata are all covered by the tag.
	header := envelope[:apiKeyEnvelopeHeaderFixedBytes+metadataLength]
	plaintext, err := HpkeOpen(suite, recipientSecretKey, enc, apiKeyEnvelopeInfo, header, ciphertext)
	if err != nil {
		return OpenedApiKey{}, err
	}
	if err := validateApiKey(plaintext); err != nil {
		wipeBytes(plaintext)
		return OpenedApiKey{}, err
	}
	metadata, err := decodeApiKeyMetadata(metadataBytes)
	if err != nil {
		wipeBytes(plaintext)
		return OpenedApiKey{}, err
	}
	if !metadata.NotAfter.IsZero() && !now.Before(metadata.NotAfter) {
		wipeBytes(plaintext)
		return OpenedApiKey{}, fmt.Errorf("envelope expired at %s", metadata.NotAfter)
	}
	return OpenedApiKey{ApiKey: plaintext, Metadata: metadata}, nil
}

// ApiKeyFingerprint returns SHA-256(public key) truncated to
// ApiKeyFingerprintBytes bytes.
func ApiKeyFingerprint(publicKey []byte) ([]byte, error) {
	return apiKeyFingerprint(publicKey)
}

func apiKeyFingerprint(publicKey []byte) ([]byte, error) {
	if len(publicKey) != X25519KeyBytes {
		return nil, fmt.Errorf("public key must be %d bytes", X25519KeyBytes)
	}
	sum := sha256.Sum256(publicKey)
	return sum[:ApiKeyFingerprintBytes], nil
}

// RandomApiKey returns a fresh ApiKeyLength-character alphanumeric API key,
// uniform over the 62-character alphabet (rejection sampling — % 62 on a
// random byte would favour the first 8 characters).
func RandomApiKey() ([]byte, error) {
	return RandomApiKeyFrom(rand.Reader)
}

// RandomApiKeyFrom returns a fresh API key from a caller-supplied source of
// randomness.
func RandomApiKeyFrom(random io.Reader) ([]byte, error) {
	key := make([]byte, ApiKeyLength)
	buf := make([]byte, ApiKeyLength)
	filled := 0
	for filled < ApiKeyLength {
		if _, err := io.ReadFull(random, buf); err != nil {
			return nil, err
		}
		for i := 0; i < len(buf) && filled < ApiKeyLength; i++ {
			sample := int(buf[i])
			if sample < apiKeySampleLimit {
				key[filled] = apiKeyAlphabet[sample%len(apiKeyAlphabet)]
				filled++
			}
		}
	}
	return key, nil
}

func apiKeyEnvelopeHeader(suite HpkeSuite, fingerprint, metadata []byte) ([]byte, error) {
	if len(metadata) > apiKeyEnvelopeMaxMetadataBytes {
		return nil, fmt.Errorf("envelope metadata too long: %d", len(metadata))
	}
	out := make([]byte, 0, apiKeyEnvelopeHeaderFixedBytes+len(metadata))
	out = append(out, apiKeyEnvelopeMagic...)
	out = binary.BigEndian.AppendUint16(out, HpkeKemID)
	out = binary.BigEndian.AppendUint16(out, HpkeKdfID)
	out = binary.BigEndian.AppendUint16(out, uint16(suite.aeadID))
	out = append(out, fingerprint...)
	out = binary.BigEndian.AppendUint16(out, uint16(len(metadata)))
	out = append(out, metadata...)
	return out, nil
}

func validateApiKey(apiKey []byte) error {
	if len(apiKey) != ApiKeyLength {
		return fmt.Errorf("API key must be %d characters", ApiKeyLength)
	}
	for _, c := range apiKey {
		alphanumeric := (c >= '0' && c <= '9') || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
		if !alphanumeric {
			return errors.New("API key must be alphanumeric")
		}
	}
	return nil
}
